### ISSUE TITLE
[question] How to build this library on macOS

### ISSUE THREAD

**Issue by oyljerry** [Thalhammer/jwt-cpp#218]
**Describe the impediment**
I want to use this library on macOS, But I could not use it correctly.

**Trying to obtain**
I download master zip from repo, then follow below steps.
```
mkdir build && cd build
mkdir install
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=./install -DJWT_DISABLE_BASE64=ON -DJWT_DISABLE_PICOJSON=OFF -DJWT_BUILD_EXAMPLES=OFF -DOPENSSL_ROOT_DIR=/Users/xxx/openssl ..
make
make install
```

**Desktop (please complete the following information):** (Include/Attach if Applicable)
 - OS: macOS
 - Compiler: AppleClang
 - Version 12

**Additional context**
After cmake and make install, the generated files in **install** folder (jwt-cpp and picojson).
Then, I include these two folder into project, and try sample code.

```
#include <jwt-cpp/jwt.h>
#include <iostream>

std::string token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXUyJ9.eyJpc3MiOiJhdXRoMCJ9.AbIJTDMFc7yUa5MhvcP03nJPyCPzZtQcGEp-zWfOkEE";
    auto decoded = jwt::decode(token);

```
The error is as below.
```
No matching function for call to 'decode'
```

**Logs** (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```
-- The C compiler identification is AppleClang 12.0.0.12000032
-- The CXX compiler identification is AppleClang 12.0.0.12000032
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found OpenSSL: /Users/xxx/openssl/lib/libcrypto.a (found suitable version "1.1.1l", minimum required is "1.0.1")
-- Configuring done
-- Generating done
```

</details>

**Comment by prince-chrismc** [comment]
Hmmm 🤔 looks good at first glance...

Can you please share your `CMakeList.txt`? As well as the verbose build command/output? That might give us more clues 🤞

**Comment by oyljerry** [comment]
CMakeLists.txt is from source code zip. I do not change anything.
```
cmake_minimum_required(VERSION 3.12)

# HUNTER_ENABLED is always set if this package is included in a project using hunter (HunterGate sets it) In this case
# we will use hunter as well to stay consistent. If not the use can supply it on configure to force using hunter.
if(HUNTER_ENABLED)
  include("cmake/HunterGate.cmake")
  huntergate(URL "https://github.com/cpp-pm/hunter/archive/v0.23.314.tar.gz" SHA1
             "95c47c92f68edb091b5d6d18924baabe02a6962a")
  message(STATUS "jwt-cpp: using hunter for dependency resolution")
endif()

project(jwt-cpp)

option(JWT_BUILD_EXAMPLES "Configure CMake to build examples (or not)" ON)
option(JWT_BUILD_TESTS "Configure CMake to build tests (or not)" OFF)
option(JWT_ENABLE_COVERAGE "Enable code coverage testing" OFF)
option(JWT_ENABLE_FUZZING "Enable fuzz testing" OFF)

option(JWT_EXTERNAL_PICOJSON "Use find_package() to locate picojson, provided to integrate with package managers" OFF)
option(JWT_DISABLE_BASE64 "Do not include the base64 implementation from this library" OFF)
option(JWT_DISABLE_PICOJSON "Do not provide the picojson template specialiaze" OFF)

set(JWT_SSL_LIBRARY_OPTIONS OpenSSL LibreSSL wolfSSL)
set(JWT_SSL_LIBRARY OpenSSL CACHE STRING "Determines which SSL library to build with")
set_property(CACHE JWT_SSL_LIBRARY PROPERTY STRINGS ${JWT_SSL_LIBRARY_OPTIONS})

set(JWT_JSON_TRAITS_OPTIONS boost-json danielaparker-jsoncons kazuho-picojson nlohmann-json)

list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake")

if(NOT JWT_SSL_LIBRARY IN_LIST JWT_SSL_LIBRARY_OPTIONS)
  message(FATAL_ERROR "JWT_SSL_LIBRARY must be one of ${JWT_SSL_LIBRARY_OPTIONS}")
endif()

# If Hunter is enabled, we configure it to resolve OpenSSL and warn the user if he selected an option not supported by
# hunter. We fall back to the system library in this case.
if(HUNTER_ENABLED)
  if(${JWT_SSL_LIBRARY} MATCHES "OpenSSL")
    hunter_add_package(OpenSSL)
  elseif(${JWT_SSL_LIBRARY} MATCHES "LibreSSL")
    message(WARNING "Hunter does not support LibreSSL yet, the system library will be used (if available)")
  elseif(${JWT_SSL_LIBRARY} MATCHES "wolfSSL")
    message(WARNING "Hunter does not support wolfSSL yet, the system library will be used (if available)")
  endif()
  if(JWT_EXTERNAL_PICOJSON)
    message(WARNING "Hunter does not support picojson yet, the system library will be used (if available)")
  endif()
endif()

# Lookup dependencies
if(${JWT_SSL_LIBRARY} MATCHES "OpenSSL")
  find_package(OpenSSL 1.0.1 REQUIRED)
elseif(${JWT_SSL_LIBRARY} MATCHES "LibreSSL")
  find_package(LibreSSL 3.0.0 REQUIRED)
elseif(${JWT_SSL_LIBRARY} MATCHES "wolfSSL")
  find_package(PkgConfig REQUIRED)
  pkg_check_modules(wolfssl REQUIRED IMPORTED_TARGET wolfssl)
  list(TRANSFORM wolfssl_INCLUDE_DIRS APPEND "/wolfssl") # This is required to access OpenSSL compatability API
endif()

if(JWT_EXTERNAL_PICOJSON)
  find_package(picojson 1.3.0 REQUIRED)
endif()

set(JWT_INCLUDE_PATH ${CMAKE_CURRENT_SOURCE_DIR}/include)
set(JWT_HEADER_FILES ${JWT_INCLUDE_PATH}/jwt-cpp/jwt.h)
foreach(traits ${JWT_JSON_TRAITS_OPTIONS})
  list(APPEND JWT_HEADER_FILES ${JWT_INCLUDE_PATH}/jwt-cpp/traits/${traits}/defaults.h
       ${JWT_INCLUDE_PATH}/jwt-cpp/traits/${traits}/traits.h)
endforeach()

if(NOT JWT_DISABLE_BASE64)
  list(APPEND JWT_HEADER_FILES ${JWT_INCLUDE_PATH}/jwt-cpp/base.h)
endif()

add_library(jwt-cpp INTERFACE)
add_library(jwt-cpp::jwt-cpp ALIAS jwt-cpp) # To match export
target_compile_features(jwt-cpp INTERFACE cxx_std_11)
if(JWT_DISABLE_BASE64)
  target_compile_definitions(jwt-cpp INTERFACE JWT_DISABLE_BASE64)
endif()
if(JWT_DISABLE_PICOJSON)
  target_compile_definitions(jwt-cpp INTERFACE JWT_DISABLE_PICOJSON)
endif()

include(GNUInstallDirs)
include(CMakePackageConfigHelpers)
target_include_directories(jwt-cpp INTERFACE $<BUILD_INTERFACE:${JWT_INCLUDE_PATH}>
                                             $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>)

if(${JWT_SSL_LIBRARY} MATCHES "OpenSSL")
  target_link_libraries(jwt-cpp INTERFACE OpenSSL::SSL OpenSSL::Crypto)
endif()

if(${JWT_SSL_LIBRARY} MATCHES "LibreSSL")
  target_link_libraries(jwt-cpp INTERFACE LibreSSL::TLS)
endif()

if(${JWT_SSL_LIBRARY} MATCHES "wolfSSL")
  target_link_libraries(jwt-cpp INTERFACE PkgConfig::wolfssl)
  # This is required to access OpenSSL compatability API
  target_include_directories(jwt-cpp INTERFACE ${wolfssl_INCLUDE_DIRS})
  target_compile_definitions(jwt-cpp INTERFACE OPENSSL_EXTRA OPENSSL_ALL)
endif()

if(JWT_EXTERNAL_PICOJSON)
  target_link_libraries(jwt-cpp INTERFACE picojson::picojson>)
endif()

# Hunter needs relative paths so the files are palced correctly
if(NOT JWT_CMAKE_FILES_INSTALL_DIR)
  set(JWT_CMAKE_FILES_INSTALL_DIR cmake)
endif()

configure_package_config_file(
  ${CMAKE_CURRENT_LIST_DIR}/cmake/jwt-cpp-config.cmake.in ${CMAKE_CURRENT_BINARY_DIR}/jwt-cpp-config.cmake
  INSTALL_DESTINATION ${JWT_CMAKE_FILES_INSTALL_DIR} PATH_VARS JWT_EXTERNAL_PICOJSON JWT_SSL_LIBRARY)
write_basic_package_version_file(${CMAKE_CURRENT_BINARY_DIR}/jwt-cpp-config-version.cmake VERSION 0.6.0
                                 COMPATIBILITY ExactVersion)

install(TARGETS jwt-cpp EXPORT jwt-cpp-targets PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
install(EXPORT jwt-cpp-targets NAMESPACE jwt-cpp:: FILE jwt-cpp-targets.cmake
        DESTINATION ${JWT_CMAKE_FILES_INSTALL_DIR})
install(DIRECTORY ${JWT_INCLUDE_PATH}/jwt-cpp DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
if(NOT JWT_EXTERNAL_PICOJSON AND NOT JWT_DISABLE_PICOJSON)
  install(FILES ${JWT_INCLUDE_PATH}/picojson/picojson.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/picojson)
endif()
install(FILES ${CMAKE_CURRENT_BINARY_DIR}/jwt-cpp-config.cmake ${CMAKE_CURRENT_BINARY_DIR}/jwt-cpp-config-version.cmake
        DESTINATION ${JWT_CMAKE_FILES_INSTALL_DIR})

if(JWT_BUILD_EXAMPLES)
  add_subdirectory(example)
endif()

if(JWT_BUILD_TESTS)
  add_subdirectory(tests)
endif()

if(JWT_ENABLE_FUZZING)
  add_subdirectory(tests/fuzz)
endif()
```

**Comment by oyljerry** [comment]
Below is make install output

```
$ make install
Install the project...
-- Install configuration: "Release"
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/cmake/jwt-cpp-targets.cmake
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/.DS_Store
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/kazuho-picojson
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/kazuho-picojson/traits.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/kazuho-picojson/defaults.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/nlohmann-json
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/nlohmann-json/traits.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/nlohmann-json/defaults.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/danielaparker-jsoncons
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/danielaparker-jsoncons/traits.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/danielaparker-jsoncons/defaults.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/boost-json
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/boost-json/traits.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/boost-json/defaults.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/traits/defaults.h.mustache
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/jwt.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/jwt-cpp/base.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/include/picojson/picojson.h
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/cmake/jwt-cpp-config.cmake
-- Installing: /XXX//jwt-cpp-0.6.0/build/install/cmake/jwt-cpp-config-version.cmake

```

**Comment by oyljerry** [comment]
This is generated Makefile

```
# CMAKE generated file: DO NOT EDIT!
# Generated by "Unix Makefiles" Generator, CMake Version 3.22

# Default target executed when no arguments are given to make.
default_target: all
.PHONY : default_target

# Allow only one "make -f Makefile2" at a time, but pass parallelism.
.NOTPARALLEL:

#=============================================================================
# Special targets provided by cmake.

# Disable implicit rules so canonical targets will work.
.SUFFIXES:

# Disable VCS-based implicit rules.
% : %,v

# Disable VCS-based implicit rules.
% : RCS/%

# Disable VCS-based implicit rules.
% : RCS/%,v

# Disable VCS-based implicit rules.
% : SCCS/s.%

# Disable VCS-based implicit rules.
% : s.%

.SUFFIXES: .hpux_make_needs_suffix_list

# Command-line flag to silence nested $(MAKE).
$(VERBOSE)MAKESILENT = -s

#Suppress display of executed commands.
$(VERBOSE).SILENT:

# A target that is always out of date.
cmake_force:
.PHONY : cmake_force

#=============================================================================
# Set environment variables for the build.

# The shell in which to execute make rules.
SHELL = /bin/sh

# The CMake executable.
CMAKE_COMMAND = /usr/local/Cellar/cmake/3.22.3/bin/cmake

# The command to remove a file.
RM = /usr/local/Cellar/cmake/3.22.3/bin/cmake -E rm -f

# Escaping for special characters.
EQUALS = =

# The top-level source directory on which CMake was run.
CMAKE_SOURCE_DIR = /Users/XXX/Downloads/jwt-cpp-0.6.0

# The top-level build directory on which CMake was run.
CMAKE_BINARY_DIR = /Users/XXX/Downloads/jwt-cpp-0.6.0/build

#=============================================================================
# Targets provided globally by CMake.

# Special rule for the target edit_cache
edit_cache:
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Running CMake cache editor..."
	/usr/local/Cellar/cmake/3.22.3/bin/ccmake -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR)
.PHONY : edit_cache

# Special rule for the target edit_cache
edit_cache/fast: edit_cache
.PHONY : edit_cache/fast

# Special rule for the target rebuild_cache
rebuild_cache:
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Running CMake to regenerate build system..."
	/usr/local/Cellar/cmake/3.22.3/bin/cmake --regenerate-during-build -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR)
.PHONY : rebuild_cache

# Special rule for the target rebuild_cache
rebuild_cache/fast: rebuild_cache
.PHONY : rebuild_cache/fast

# Special rule for the target list_install_components
list_install_components:
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Available install components are: \"Unspecified\""
.PHONY : list_install_components

# Special rule for the target list_install_components
list_install_components/fast: list_install_components
.PHONY : list_install_components/fast

# Special rule for the target install
install: preinstall
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Install the project..."
	/usr/local/Cellar/cmake/3.22.3/bin/cmake -P cmake_install.cmake
.PHONY : install

# Special rule for the target install
install/fast: preinstall/fast
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Install the project..."
	/usr/local/Cellar/cmake/3.22.3/bin/cmake -P cmake_install.cmake
.PHONY : install/fast

# Special rule for the target install/local
install/local: preinstall
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing only the local directory..."
	/usr/local/Cellar/cmake/3.22.3/bin/cmake -DCMAKE_INSTALL_LOCAL_ONLY=1 -P cmake_install.cmake
.PHONY : install/local

# Special rule for the target install/local
install/local/fast: preinstall/fast
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing only the local directory..."
	/usr/local/Cellar/cmake/3.22.3/bin/cmake -DCMAKE_INSTALL_LOCAL_ONLY=1 -P cmake_install.cmake
.PHONY : install/local/fast

# Special rule for the target install/strip
install/strip: preinstall
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing the project stripped..."
	/usr/local/Cellar/cmake/3.22.3/bin/cmake -DCMAKE_INSTALL_DO_STRIP=1 -P cmake_install.cmake
.PHONY : install/strip

# Special rule for the target install/strip
install/strip/fast: preinstall/fast
	@$(CMAKE_COMMAND) -E cmake_echo_color --switch=$(COLOR) --cyan "Installing the project stripped..."
	/usr/local/Cellar/cmake/3.22.3/bin/cmake -DCMAKE_INSTALL_DO_STRIP=1 -P cmake_install.cmake
.PHONY : install/strip/fast

# The main all target
all: cmake_check_build_system
	$(CMAKE_COMMAND) -E cmake_progress_start /Users/XXX/Downloads/jwt-cpp-0.6.0/build/CMakeFiles /Users/XXX/Downloads/jwt-cpp-0.6.0/build//CMakeFiles/progress.marks
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 all
	$(CMAKE_COMMAND) -E cmake_progress_start /Users/XXX/Downloads/jwt-cpp-0.6.0/build/CMakeFiles 0
.PHONY : all

# The main clean target
clean:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 clean
.PHONY : clean

# The main clean target
clean/fast: clean
.PHONY : clean/fast

# Prepare targets for installation.
preinstall: all
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 preinstall
.PHONY : preinstall

# Prepare targets for installation.
preinstall/fast:
	$(MAKE) $(MAKESILENT) -f CMakeFiles/Makefile2 preinstall
.PHONY : preinstall/fast

# clear depends
depend:
	$(CMAKE_COMMAND) -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR) --check-build-system CMakeFiles/Makefile.cmake 1
.PHONY : depend

# Help Target
help:
	@echo "The following are some of the valid targets for this Makefile:"
	@echo "... all (the default if no target is provided)"
	@echo "... clean"
	@echo "... depend"
	@echo "... edit_cache"
	@echo "... install"
	@echo "... install/local"
	@echo "... install/strip"
	@echo "... list_install_components"
	@echo "... rebuild_cache"
.PHONY : help



#=============================================================================
# Special targets to cleanup operation of make.

# Special rule to run CMake to check the build system integrity.
# No rule that depends on this can have commands that come from listfiles
# because they might be regenerated.
cmake_check_build_system:
	$(CMAKE_COMMAND) -S$(CMAKE_SOURCE_DIR) -B$(CMAKE_BINARY_DIR) --check-build-system CMakeFiles/Makefile.cmake 0
.PHONY : cmake_check_build_system


```

**Comment by prince-chrismc** [comment]
I appogize, my question was not clear. Let me try again!

I wanted to know how are you building the example code? I am curious how you are getting the compilation error. What build command are you using to get it?

You posted CMake logs so I assumed (incorrectly) you were building it separately. Also you have `EXAMPLES=OFF` in the configure which further lead me astray.

**Comment by oyljerry** [comment]
Sorry, I made a mistake.
When I compile, I disable DJWT_DISABLE_BASE64, so when decode, It shows **No matching function for call to 'decode'**.

**Comment by prince-chrismc** [comment]
I am glad you solved your issue!

If there is anything you can suggest to improve the example please let us know, always looking to improve ❤️
